### PR TITLE
new_alpha_3_4_pipeline

### PR DESCRIPTION
**Pull request by pablo-lauret**
## Scope
This PR adds the working pipeline for Alpha 3/4. There is one thing that may need to be changed, but it is a small change and we will discuss it together in the meeting later - this change can be pushed later and a new PR can be made. Pipeline running smoothly with no errors on a subset of the data to be used in the experiments.

A tiny change was made to the alpha_3_4 module (doesn't break any tests) to ensure labels are fed to the loss function as ints and not as floats - this was causing a strange error that is now fixed.

## Emoji Guide
**For reviewers: Emojis can be added to comments to call out blocking versus non-blocking feedback.**
E.g: Praise, minor suggestions, or clarifying questions that don’t block merging the PR.
> 🟢 Nice refactor!
> 🟡 Why was the default value removed?
E.g: Blocking feedback must be addressed before merging.
> 🔴 This change will break something important

| | | |
| --- | --- | --- |
| Blocking | 🔴 ❌ 🚨 | Red |
| Non-blocking | 🟡 💡 🤔 💭 | Yellow, thinking, etc |
| Praise | 🟢 💚 😍 👍 🙌 | Green, hearts, positive emojis, etc |

### CODE REVIEW

File: neasqc_wp61/data/data_processing/use_alpha_3_4.py
Comment (ryk-wolf) [inline neasqc_wp61/data/data_processing/use_alpha_3_4.py:1]: I am uncomfortable having default arguments for everything.
I don't think I can see a use case right now where it makes sense to have a default, if we haven't provided that argument, it means we're doing the XP wrong, so I'd much rather have it crash / raise an error than go about it silently if we've messed up.

Comment (ryk-wolf) [reply neasqc_wp61/data/data_processing/use_alpha_3_4.py:1]: Let's remove the "default" values in the parser arguments

File: neasqc_wp61/data/data_processing/use_alpha_3_4.py [hunk 1]
```diff
@@ -230,167 +221,204 @@ def main(args) -> dict:
...
 
-    split_data_path = dataset_path.split('_')
-    split_data_path = [item.split('.') for item in split_data_path]
-
-    if args.embedder not in split_data_path:
-            raise ValueError("The dataset provided does not correspond"
-                             " to the choice of embedder.")
+    if args.embedder.lower() not in dataset_path:
+        raise ValueError(
+            "The dataset provided does not correspond"
```
Comment (ryk-wolf) [inline neasqc_wp61/data/data_processing/use_alpha_3_4.py:286]: I don't understand this error message: what's the link between dataset and embedder here?

File: neasqc_wp61/data/data_processing/use_alpha_3_4.py [hunk 2]
```diff
@@ -230,167 +221,204 @@ def main(args) -> dict:
...
+            " to the choice of embedder."
+        )
 
     dataset = pd.read_pickle(dataset_path)
     dim_red.fit(data_to_fit=dataset)
```
Comment (ryk-wolf) [inline neasqc_wp61/data/data_processing/use_alpha_3_4.py:291]: Are we sure this is correct? we're fitting over whole dataset??? Looks strange to me right now. Pls check

File: neasqc_wp61/data/data_processing/use_alpha_3_4.py [hunk 3]
```diff
@@ -230,167 +221,204 @@ def main(args) -> dict:
...
 
     # Split train and val
-    val = dataset[dataset['split'] == split_id]
-    train = dataset[dataset['split'] != split_id]
+    val = dataset[dataset["split"] == split_id]
```
Comment (ryk-wolf) [inline neasqc_wp61/data/data_processing/use_alpha_3_4.py:294]: In your comments earlier you said you kept the ID+1 to be test. This doesn't seem to be what you're doing here.
Pls double-check and make sur you're doing what you want to and should be doing.

File: neasqc_wp61/data/data_processing/use_alpha_3_4.py
Comment (ryk-wolf) [reply neasqc_wp61/data/data_processing/use_alpha_3_4.py:1]: Let's make sure we're removed all the debugging "print" statements. No use keeping them in the HPC pipeline.

File: neasqc_wp61/data/data_processing/use_alpha_3_4.py [hunk 4]
```diff
@@ -230,167 +221,204 @@ def main(args) -> dict:
...
+    print(
+        f"----------\n"
+        "Beginning training for the set of parameters with ID: {args.exp_id}\n"
+        "----------\n"
+    )"""
     t_before = time.time()
     model.train()
     t_after = time.time()
-    print('\n----- Training completed! -----\n')
+    # print("\n----- Training completed! -----\n")
```
Comment (ryk-wolf) [inline neasqc_wp61/data/data_processing/use_alpha_3_4.py]: remove?

File: neasqc_wp61/data/data_processing/use_alpha_3_4.py [hunk 5]
```diff
@@ -230,167 +221,204 @@ def main(args) -> dict:
...
     train_preds_list = model.preds_train
     train_probs_list = model.probs_train
-    val_loss_list =  model.loss_val
+    val_loss_list = model.loss_val
     val_preds_list = model.preds_val
     val_probs_list = model.probs_val
 
     time_taken = t_after - t_before
-    print(f"Time taken to run ID {args.exp_id} = {time_taken}\n")
+    # print(f"Time taken to run = {time_taken}\n")
```
Comment (ryk-wolf) [inline neasqc_wp61/data/data_processing/use_alpha_3_4.py]: remove?

File: neasqc_wp61/data/data_processing/use_alpha_3_4.py [hunk 6]
```diff
@@ -230,167 +221,204 @@ def main(args) -> dict:
...
+    )"""
     t_before = time.time()
     model.train()
     t_after = time.time()
-    print('\n----- Training completed! -----\n')
+    # print("\n----- Training completed! -----\n")
 
     train_loss_list = model.loss_train
     train_preds_list = model.preds_train
     train_probs_list = model.probs_train
```
Comment (ryk-wolf) [inline neasqc_wp61/data/data_processing/use_alpha_3_4.py:360]: What does this correpsond to? I see line 347: output probabilities = False so are we happy to have output probs here?

File: neasqc_wp61/data/data_processing/use_alpha_3_4.py [hunk 7]
```diff
@@ -230,167 +221,204 @@ def main(args) -> dict:
...
 
     # Train model and store model outputs
-    
-    model.seed = args.seed
-
-    print(f'----------\n /
-           Beginning training for the set of parameters with ID: {args.exp_id}\n /
-           ----------\n')
+    """
+    print(
```
Comment (ryk-wolf) [inline neasqc_wp61/data/data_processing/use_alpha_3_4.py]: remove

File: neasqc_wp61/models/quantum/alpha_3_4/alpha_3_4.py [hunk 8]
```diff
@@ -190,8 +192,12 @@ def train(self) -> None:
                     vector = vector.to(self.device)
                     label = label.to(self.device)
                     opt.zero_grad()
+                    print(self.forward(vector))
```
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/alpha_3_4/alpha_3_4.py]: do we needs those prints???

**Review by ryk-wolf — CHANGES_REQUESTED** [review]
Great job! Loads of careful work went into this, it's good that we're nearing the end and can move on to XPs!!!!


Pls make extra sure to double check the logic on the points I flagged

We can all discuss together at meeting if still any doubt abt them.

Also do make sure you're up to date with dev so there are no merging conflicts.

and do remove debug prints :,)

File: neasqc_wp61/data/data_processing/use_alpha_3_4.py (hunk 1 above)
Comment (pablo-lauret) [reply neasqc_wp61/data/data_processing/use_alpha_3_4.py:286]: Hi Ryk. Given that we are not using the embedder, and that we are providing the vectorised datasets to the pipeline to accelerate the workflow, in this current state, the embedder serves as a sanity check that the dataset provided has indeed been previously vectorised with that embedder. I can remove the embedder argument altogether if you prefer. Let me know

File: neasqc_wp61/data/data_processing/use_alpha_3_4.py (hunk 2 above)
Comment (pablo-lauret) [reply neasqc_wp61/data/data_processing/use_alpha_3_4.py:291]: In the end Bruno joined only train + val. Test will be separate, i.e. the test file given by Tilde in their repo. This is to prevent the model from ever seeing any test data. Given that testing will be done by a separate script and on a separate dataset, I removed it. Now the full dataset is train+val with 10-fold cross-validation

File: neasqc_wp61/data/data_processing/use_alpha_3_4.py (hunk 3 above)
Comment (pablo-lauret) [reply neasqc_wp61/data/data_processing/use_alpha_3_4.py:294]: please see reply to above comment

File: neasqc_wp61/data/data_processing/use_alpha_3_4.py (hunk 6 above)
Comment (pablo-lauret) [reply neasqc_wp61/data/data_processing/use_alpha_3_4.py:360]: The output_probabilities attribute is of the Circuit class and has to do with the probabilities of measurements of the quantum circuit, it is True for Alpha4 and False for Alpha3. It is different to the NN's probabilities, which is the vector from which the final prediction is made. We need these to calculate Precision, Recall, Accuracy, etc. of the model.

File: neasqc_wp61/models/quantum/alpha_3_4/alpha_3_4.py (hunk 8 above)
Comment (pablo-lauret) [reply neasqc_wp61/models/quantum/alpha_3_4/alpha_3_4.py]: nope this was for debugging, removing

File: neasqc_wp61/data/data_processing/use_alpha_3_4.py [hunk 9]
```diff
@@ -0,0 +1,401 @@
...
+
+    # Prepare data
+
+    # Set dataset
+    if args.dataset == "ag_news":
+        dataset_path = "../datasets/ag_news_bert_balanced_cvsplit.pkl"
+        n_classes = 4
+    elif args.dataset == "newscatcher":
+        n_classes = 7
+        pass
```
Comment (ryk-wolf) [inline neasqc_wp61/data/data_processing/use_alpha_3_4.py:269]: remove pass, it does nothing here

Comment (ryk-wolf) [reply neasqc_wp61/data/data_processing/use_alpha_3_4.py:269]: Also how come this one doesn't have a path to the dataset?

File: neasqc_wp61/data/data_processing/use_alpha_3_4.py [hunk 10]
```diff
@@ -0,0 +1,401 @@
...
+    elif args.dataset == "reviews":
+        n_classes = 3
+        pass
```
Comment (ryk-wolf) [inline neasqc_wp61/data/data_processing/use_alpha_3_4.py:272]: SAA

File: neasqc_wp61/data/data_processing/use_alpha_3_4.py (hunk 1 above)
Comment (ryk-wolf) [reply neasqc_wp61/data/data_processing/use_alpha_3_4.py:286]: If we're not using it here to embed stuff, pls remove it. No unnecessary code trailing around.

File: neasqc_wp61/data/data_processing/use_alpha_3_4.py (hunk 2 above)
Comment (ryk-wolf) [reply neasqc_wp61/data/data_processing/use_alpha_3_4.py:291]: great

File: neasqc_wp61/data/data_processing/use_alpha_3_4.py [hunk 11]
```diff
@@ -0,0 +1,401 @@
...
+        ].tolist()
+        sentence_vectors.append(reduced_embeddings)
+
+    # Initialise model with all of its attributes
+    if args.model == "alpha_3":
+        circuit = ansatz(
+            n_qubits=args.n_qubits,
+            n_layers=args.n_layers,
+            axis_embedding="X",
+            observables={key: qml.PauliZ for key in range(args.n_qubits)},
```
Comment (ryk-wolf) [inline neasqc_wp61/data/data_processing/use_alpha_3_4.py:313]: observable PauliZ?

**Review by ryk-wolf — CHANGES_REQUESTED** [review]
Apart from the question about PauliZ observable, everything is pretty minor.
Amazing  job you've been doing! Keep it up 💪

File: neasqc_wp61/data/data_processing/use_alpha_3_4.py (hunk 11 above)
Comment (ryk-wolf) [reply neasqc_wp61/data/data_processing/use_alpha_3_4.py:313]: should be Y

**Review by ryk-wolf — APPROVED** [review]
